### PR TITLE
Add new driver specific configurations

### PR DESCRIPTION
**Pull request by Madhu-1**
This PR does below changes

* Add New DriverConfig structure which holds the required information to establish the connection with CSI driver
* dynamic generation of  leader election ID 
* Add leader election Namespace
* Change health-probe-bind-address to `9998`
* Disable Metrics by default

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

### CODE REVIEW

**Review by iamniting — APPROVED** [review]

File: controllers/volumereplication_controller.go [hunk 1]
```diff
@@ -25,13 +25,15 @@ import (
...
 	replicationv1alpha1 "github.com/kube-storage/volume-replication-operator/api/v1alpha1"
+	"github.com/kube-storage/volume-replication-operator/pkg/config"
 )
 
 // VolumeReplicationReconciler reconciles a VolumeReplication object
 type VolumeReplicationReconciler struct {
 	client.Client
 	Log    logr.Logger
 	Scheme *runtime.Scheme
+	Config *config.DriverConfig
```
Comment (umangachapagain) [inline controllers/volumereplication_controller.go:36]:
```suggestion
	DriverConfig *config.DriverConfig
```
to be clearer?

File: main.go [hunk 2]
```diff
@@ -50,10 +58,18 @@ func init() {
...
 	var probeAddr string
-	flag.StringVar(&metricsAddr, "metrics-bind-address", ":8080", "The address the metric endpoint binds to.")
-	flag.StringVar(&probeAddr, "health-probe-bind-address", ":8081", "The address the probe endpoint binds to.")
+
+	cfg := config.NewDriverConfig()
+
+	flag.StringVar(&metricsAddr, "metrics-bind-address", "0", "The address the metric endpoint binds to.")
+	flag.StringVar(&leaderElectionNamespace, "leader-election-namespace", "", "Namespace where the leader election resource lives")
+	flag.StringVar(&cfg.DriverName, "driver-name", "", "The CSI driver name.")
+	flag.StringVar(&cfg.DriverEndPoint, "csi-address", "/run/csi/socket", "Address of the CSI driver socket.")
```
Comment (umangachapagain) [inline main.go:70]:
```suggestion
	flag.StringVar(&cfg.DriverEndpoint, "csi-address", "/run/csi/socket", "Address of the CSI driver socket.")
```

File: main.go [hunk 3]
```diff
@@ -50,10 +58,18 @@ func init() {
...
+	flag.DurationVar(&cfg.Timeout, "timeout", defaultTimeout, "The timeout for RPCs to the CSI driver.")
```
Comment (umangachapagain) [inline main.go:71]:
```suggestion
	flag.DurationVar(&cfg.Timeout, "rpc-timeout", defaultTimeout, "The timeout for RPCs to the CSI driver.")
```

File: main.go [hunk 4]
```diff
@@ -50,10 +58,18 @@ func init() {
...
+	flag.StringVar(&probeAddr, "health-probe-bind-address", ":9998", "The address the probe endpoint binds to.")
```
Comment (umangachapagain) [inline main.go:72]: Should we keep this as before and change it in our deployment configuration? Or do you want to enforce this default to everyone (unless they want to change it) ?

File: main.go [hunk 5]
```diff
@@ -65,13 +81,28 @@ func main() {
...
 	ctrl.SetLogger(zap.New(zap.UseFlagOptions(&opts)))
 
+	err := cfg.Validate()
+	if err != nil {
+		setupLog.Error(err, "error in driver configuration")
+		os.Exit(1)
+	}
+
+	if leaderElectionNamespace == "" {
+		fmt.Println("leader-election-namespace is empty")
```
Comment (umangachapagain) [inline main.go:91]: Should/Can this be logged as well?

File: pkg/config/config.go [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+import (
+	"errors"
+	"time"
+)
+
+// DriverConfig holds the CSI driver configuration.
+type DriverConfig struct {
+	// The gRPC endpoint for CSI driver.
+	DriverEndPoint string
```
Comment (umangachapagain) [inline pkg/config/config.go:27]:
```suggestion
	DriverEndpoint string
```

File: pkg/config/config.go [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+	// The name of CSI Driver.
+	DriverName string
+	// The timeout for RPCs to the CSI driver.
+	Timeout time.Duration
```
Comment (umangachapagain) [inline pkg/config/config.go:31]:
```suggestion
	RPCTimeout time.Duration
```

**Review by sp98 — COMMENTED** [review]
nit. 
rest LGTM

**Review by sp98 — APPROVED** [review]

File: main.go (hunk 4 above)
Comment (Madhu-1) [reply main.go:72]: changed it 9998 as 8081 is the most common and widely used one.

File: main.go (hunk 5 above)
Comment (Madhu-1) [reply main.go:91]: I tried to log initially it becomes messy as we need to define a new error variable and also `setupLog.Error` Needs two arguments.

File: main.go (hunk 4 above)
Comment (umangachapagain) [reply main.go:72]: I had some other thoughts, but I this is fine too.

File: main.go (hunk 5 above)
Comment (umangachapagain) [reply main.go:91]: yes, but with this approach we lose log when we exit. is there a way to debug why we exited?
If yes, then I won't insist on logging.

Comment (Madhu-1) [reply main.go]: done now logging to stderr

**Review by umangachapagain — APPROVED** [review]

**Pull request merged**